Richard homework

ApogeeOfLabor opened the pull request:
Прочитал заранее методичку.
Сделал все дз.

Commit:
Task2 и task3 готовы

Сделал сразу по условию со * из task3

Commit:
Task4 закончил.
Оставил принты на случай запуска проверки.

Commit:
Task5 закончил.
Оставил принты с проверками условий.

Commit:
Закончил 1 и 2 задание.

Commit:
Внёс не большие исправления по заданию и постарался оптимизировать алгоритм

Commit:
Task3 Done/

Commit:
Task4 Lesson 3 в процессе.
Промежуточный комит, работаю с разных машин.

Commit:
Task4 Done

Commit:
Исправляем код под условие в задании.

Commit:
task5 в процессе

Commit:
task5 в процессе (смена машины)

Commit:
task5 завершено

Commit:
Инициализация каталогов для будущих дз.

Commit:
Инициализация каталогов для будущих дз.

Commit:
ReFormat кода

Commit:
ReFormat кода

Commit:
Lesson4 task1.py - Done.

Commit:
Merge remote-tracking branch 'origin/richard-hw' into richard-hw

Commit:
Lesson4 Task2 Done.

pimenovdv commented on Ehrlich_Richard_lesson2/task2_and_3.py:
```diff
@@ -0,0 +1,29 @@
+# Проще пока что-то не вышло)
+
+weather_report = ['в', '5', 'часов', '17', 'минут', 'температура', 'воздуха', 'была', '+5', 'градусов']
+print(id(weather_report), *weather_report)
```
Эти строки тоже лучше в main перенести (считай, что ты пишешь модуль из функций, а всё, что делает main - это просто проверка этих модулей)

ApogeeOfLabor replied:
понял) принял)

pimenovdv commented on Ehrlich_Richard_lesson2/task2_and_3.py:
```diff
@@ -0,0 +1,29 @@
+
+
+if __name__ == '__main__':
+    list_of_found_numbers = [item for item in weather_report if item[-1].isdigit()]
```
Хитро :) в целом, если бы в задаче не было "Дан список", то подразумевалось бы, что число может быть внутри текст, но здесь задача стоит чётко :) тут с одной стороны, программист должен учиться предвидеть ситуации и наперёд писать, а с другой - чётко выполнять ТЗ с оптимальной реализацией, у тебя второй вариант удался, поэтому  всё ок :) (но на будущее, учти, что если тебе говорит условный заказчик, что "дан список" - это значит, что приблизительно и такие моменты уточняй на этапе постановки задачи)

ApogeeOfLabor replied:
Если "предвидеть на перёд" имелось ввиду универсальность алгоритма, то да  :) я это предвидел и не стал заморачиваться )) на универсальность и сделал просто по заданию

pimenovdv commented on Ehrlich_Richard_lesson2/task2_and_3.py:
```diff
@@ -0,0 +1,29 @@
+if __name__ == '__main__':
+    list_of_found_numbers = [item for item in weather_report if item[-1].isdigit()]
+    for num in list_of_found_numbers:
+        insert_quotation_mark(weather_report.index(num), weather_report)
```
этот момент естественно не по уровню, но тебе просто уточню - метод index пробегает по списку для поиска (сложность O(n)), поэтому конкретно здесь было бы лучше в list_of_found_numbers сохранять индексы таким образом:
list_of_found_numbers = [ndx for item in enumerate(weather_report) if item[-1].isdigit()]
В том варианте, как у тебя - будет медленно работать, если список большой :) ну это тоже не недочёт, а совет на будущее :)

ApogeeOfLabor replied:
про сложность алгоритмов ещё очень мало знаю, но на заметку взял , спасибо)

pimenovdv commented on Ehrlich_Richard_lesson2/task4.py:
```diff
@@ -0,0 +1,7 @@
+employee_data = ['инженер-конструктор Игорь', 'главный бухгалтер МАРИНА', 'токарь высшего разряда нИКОЛАй', 'директор аэлита']
+print(id(employee_data), employee_data)
+for item, name in zip(employee_data, [s[s.rindex(' ') + 1:].capitalize() for s in employee_data]):
+    employee_data[employee_data.index(item)] = f"{item[:item.rindex(' ')]} {name}"
```
тут вместо index можно было enumerate брать, тогда выглядеть будет выражение так:
for ndx, (item, name) in enumerate(zip...
Я так понимаю, проблема могла быть с тем, что после for нужны скобки (у меня была такая проблема раньше, теперь я знаю, что скобками можно разделять, иерархию распаковки)

ApogeeOfLabor replied:
вау) пока перевариваю ) , но очень интересная мысль
надо запомнить!  :)

pimenovdv commented on Ehrlich_Richard_lesson2/task4.py:
```diff
@@ -0,0 +1,7 @@
+employee_data = ['инженер-конструктор Игорь', 'главный бухгалтер МАРИНА', 'токарь высшего разряда нИКОЛАй', 'директор аэлита']
+print(id(employee_data), employee_data)
+for item, name in zip(employee_data, [s[s.rindex(' ') + 1:].capitalize() for s in employee_data]):
```
да, тут круто сделал, ещё посмотри такой код:
print(type([s[s.rindex(' ') + 1:].capitalize() for s in employee_data]))
print(type((s[s.rindex(' ') + 1:].capitalize() for s in employee_data)))
и подумай, как поменяв 2 символа ещё снизить потребление по памяти (по скорости не снизится). Но это уже размышления на тему генераторов :)

ApogeeOfLabor replied:
Пока не добрался) Но непременно изучу)

ApogeeOfLabor replied:
Добрался и проверил) с помощью sys.getsizeof() - значит я всё правильно делал или нет?
<class 'list'> 88
<class 'generator'> 112
Алгоритм одинаковый O(n). Возможно есть обратная разница на больших данных, но я пока не в курсе)

pimenovdv commented on Ehrlich_Richard_lesson2/task5.py:
```diff
@@ -0,0 +1,32 @@
+import random
+
+# Получаем случайный список цен (я понимаю, что в задании написано создать в ручную), но так интереснее :)
+random_number_list = [float('{:.02f}'.format(random.randint(1, 16) * num)) for num in [random.random() for _ in range(15)]]
```
перемудрил :) random.random() * 16 даст в среднем такой же результат :) там равномерное распределение вероятности :)

ApogeeOfLabor replied:
Проверил , понял ,учёл)) Спасибо!)

pimenovdv commented on Ehrlich_Richard_lesson2/task5.py:
```diff
@@ -0,0 +1,32 @@
+
+def get_price_template(price_list):
+    print_list = []
+    for item in [str(n).split('.') for n in price_list]:
```
вообще, код, конечно короче, но внутри for выполнить операцию, как по мне красивее, но это вопрос вкуса, так тоже норм :) единственное, скобки круглые нужны тут :)

ApogeeOfLabor replied:
Круглые скобки, чтобы получить генератор я так понимаю.
Правильно ли я понимаю, что (генератор) в конечном итоге возвращает генератор, что в свою очередь является по сути итератором и в конечном итоге никакой список не создаётся, а компрехеншон лист возвращает в итоге лист, что в свою очередь просто создаётся дополнительный список который как следствие больше весит или я что-то не так понял ?)

ApogeeOfLabor replied:
Хотя из предыдущей проверки генератор весил больше. Странно. Тогда я не понимаю , почему лучше генератор, а не компрехеншон лист  О_о )))

pimenovdv commented on Ehrlich_Richard_lesson2/task5.py:
```diff
@@ -0,0 +1,32 @@
+def get_price_template(price_list):
+    print_list = []
+    for item in [str(n).split('.') for n in price_list]:
+        print_list.append('{} руб {:<02d} коп,'.format(item[0], int(item[1])) if len(item[1]) == 1 else '{} руб {} коп,'.format(item[0], item[1]))
```
аа, что-то я не подумал, зачем ты вообще сделал операцию split :) тут лучше было бы // 1 и % 1 * 100 сделать и не нужны ifы :) А и ещё, такой совет - если строка длинна, переноси её (pycharm подскажет, где и как это сделать)

ApogeeOfLabor replied:
с математическими алгоритмами пока на вы , поэтому совет перевариваю )
уже начинаю подумывать о доп.курсе по алгебре и т.п.

pimenovdv commented on Ehrlich_Richard_lesson2/task5.py:
```diff
@@ -0,0 +1,32 @@
+get_price_template(random_number_list)
+
+# Step B
+# Сортировка списка без изменения объекта
```
тут как раз с изменением объекта, ты хотел сказать, без создания нового :)

Commit:
Lesson4 task2 Done.

Commit:
Lesson4 task3 Done.

Commit:
Доработка кода для того, чтобы была возможность импортировать как модуль.

Commit:
Lesson4 DONE.

Commit:
Lesson5 START.

Commit:
Lesson5 task1 and task2 - Done.

Commit:
Lesson5 task3 - start

Дулаю дз, то на работе, то дома.
Поэтому приходиться делать такие промежуточные комиты.

Commit:
Lesson4 исправлена дата

Commit:
Task 1 - 4 Done

Commit:
Lesson5 Done

pimenovdv commented on Ehrlich_Richard_Lesson3/task5.py:
```diff
@@ -0,0 +1,33 @@
+def get_jokes(lot, *args, flag=False):
+    list_jokes = []
+    while lot:
```
В принципе, подход рабочий и довольно необычный, но есть нюанс: иногда код программы может выполняться очень долго, т.к. рандом может часто попадать в одни и те же слова. Лучше в таком случае скопировать списки, например (именно deepcopy, не простым присваиванием) и при добавлении каждого анекдота убирать все использующиеся там слова, тогда алгоритм гарантированно отработает за lot итераций :) Но идея использовать map мне здесь очень понравилась :)

Commit:
Lesson 6 task1

Commit:
Lesson5 task3 оптимизация кода

Commit:
Lesson6 task1 and task2 Done

Commit:
Merge remote-tracking branch 'origin/richard-hw' into richard-hw

# Conflicts:
#	Ehrlich_Richard_Lesson6/task1.py

Commit:
Lesson6 task1 modify

Commit:
Lesson6 task2 оптимизация кода

Commit:
Lesson6 task3 старт

Commit:
Lesson6 task3 done

Commit:
Lesson6 task4 done

Commit:
Lesson6 task5 start

Commit:
Lesson6 task4 и 5 в процессе

pimenovdv commented on Ehrlich_Richard_Lesson4/task2_with_etree.py:
```diff
@@ -0,0 +1,24 @@
+    import requests
+    import xml.etree.ElementTree as ET
+
+    url = 'http://www.cbr.ru/scripts/XML_daily.asp'
```
В подобных задачах лучше скрывать ссылки внутри функций (ну или делать необязательным параметром). Ну, по крайней мере это справедливо для библиотек, чтобы пользователь не заморачивался, с какой ссылки ему искать там чего :)

pimenovdv commented on Ehrlich_Richard_Lesson4/task2_with_etree.py:
```diff
@@ -0,0 +1,24 @@
+        value_valute = inner_item.find('Value').text
+
+        if char_valute.lower() == name_valute.lower():
+            return '.'.join(value_valute.split(','))
```
проще, как мне кажется сделать это replace'ом :) да и оптимальнее, но подход интересный :)

pimenovdv commented on Ehrlich_Richard_Lesson4/task3_with_etree.py:
```diff
@@ -0,0 +1,17 @@
+    response = requests.get('http://www.cbr.ru/scripts/XML_daily.asp')
+    dom_data_valute = ET.fromstring(response.text)
+    date_obj = dateparser.parse(' '.join(response.headers['Date'].split()[1:4]))
+    end_ = str(date_obj).split()[0]
```
в Python есть негласное соглашение, что локальные переменные могут иметь префикс _, так что ты почти угадал :) В целом, это правило не является общим требованием какого-либо документа, я так понимаю, пошло оно от правила именования методов классов (приватные методы начинаются с _ или с __, об этом расскажу на одном из уроков по ООП)

pimenovdv commented on Ehrlich_Richard_Lesson4/task3_with_string.py:
```diff
@@ -0,0 +1,25 @@
+        if key == valute_code.upper():
+            for user_string in value:
+                if user_string.startswith('Value'):
+                    return float('.'.join(user_string[user_string.find('>') + 1: user_string.find('<')].split(','))), end_
```
Ну, собственно, вполне рабочий метод... Просто обычно такие задачи решаются готовыми парсерами (как ты решал), а если решать задачу таким образом, то вполне законно. Сложный вариант (как бы это нужно было делать, если это был бы парсер полноценный) - искать теги и по позициям этих тегов искать значения внутри них (соответственно, левая граница - index(tag) + len(tag)б правая - index(tag_closed)), потом выполнять trim() (чтобы убрать лишние пробелы) и потом уже работать с содержимым. Но это всё-таки для более универсальных работ, а когда у тебя конкретная задача, то такое решение тоже подходит :)

Commit:
Lesson6 task5 в процессе.

Commit:
Lesson6 task4 done

Commit:
Lesson6 task4 done

Commit:
Lesson6 task5 done

Commit:
Lesson6 task6 в процессе

pimenovdv commented on Ehrlich_Richard_Lesson5/task1.py:
```diff
@@ -0,0 +1,5 @@
+def odd_numbers(): yield from range(1, int(input()) + 1, 2)
```
Совет: лучше input() перенести в аргумент функции :)
def odd_numbers(num): yield from range(1, int(num) + 1, 2)
...odd_numbers(input())...

ApogeeOfLabor replied:
> Совет: лучше input() перенести в аргумент функции :) def odd_numbers(num): yield from range(1, int(num) + 1, 2) ...odd_numbers(input())...

а если input() перенести в аргументы функции , то как его потом оттуда брать для использования, я тогда не придул. Или ты имеешь ввиду сделать его именованным аргументом типа num=input() ?

pimenovdv commented on Ehrlich_Richard_Lesson5/task3.py:
```diff
@@ -0,0 +1,14 @@
+def get_tuple_range():
+    yield from zip(tutors, (klasses[index] if index < len(klasses) else None for index in range(len(tutors))))
```
Вместо
for index in range(len(tutors))
лучше
for i, _ in enumerate(tutors)

ApogeeOfLabor replied:
> Вместо for index in range(len(tutors)) лучше for i, _ in enumerate(tutors)

да да да)), соглсен ) косячёк)

pimenovdv commented on Ehrlich_Richard_Lesson5/task4.py:
```diff
@@ -0,0 +1,48 @@
+    # Timer нашёл в сети. Как он работает пока не понимаю.
+
+    print(*get_next_big_num(src), 'size: ->', getsizeof(get_next_big_num(src)), 'timer: ->',
+          Timer("get_next_big_num(src)", 'from __main__ import get_next_big_num').timeit(0),
```
Попробуй во всех местах поменять на 
Timer("get_next_big_num(src)", globals=globals()).timeit()
Тогда результат будет очевиднее :)
в аргумент globals передаёшь все объекты глобальной области видимости, а timeit() запустит 1 млн циклов. Твой вариант сейчас запускает 0 циклов :) а аргумент setup тут не обязательно :) это если бы ты хотел из другого модуля запустить

Commit:
Lesson6 task5 refactoring

Commit:
Lesson6 task4 code refactoring

Commit:
Lesson6 task6 пока закнчил

Commit:
Lesson6 task7 DONE

Commit:
Lesson5 task4 изменил код таймера

Commit:
Lesson7 task2 done

Commit:
Lesson7 task3 done

Commit:
Lesson7 done

Commit:
Lesson8 task1
